src/components/index.ts

```diff
@@ -0,0 +1,388 @@
+  const filterKeys = Object.keys(filters);
+
+  const filtersMap: Record<string, string[]> = filterKeys.reduce(
+    (acc, key) => ({
+      [key]: [],
+      ...acc,
+    }),
+    {},
+  );
+
+  for (const param of filterKeys) {
+    const operation = filters[param];
+
+    if (typeof operation === "boolean") {
+      const idx = index.indexes[param] as BooleanIndex;
+      const filteredIDs = idx[operation.toString() as keyof BooleanIndex];
+      filtersMap[param].push(...filteredIDs);
+      continue;
+    }
+
+    const operationKeys = Object.keys(operation);
+
+    if (operationKeys.length > 1) {
+      throw createError("INVALID_FILTER_OPERATION", operationKeys.length);
+    }
+
+    const operationOpt = operationKeys[0] as ComparisonOperator;
+    const operationValue = operation[operationOpt as unknown as keyof ComparisonOperator];
+
+    const AVLNode = index.indexes[param] as AVLNode<number, string[]>;
+
+    switch (operationOpt) {
+      case "gt": {
+        const filteredIDs = avlGreaterThan(AVLNode, operationValue, false);
+        filtersMap[param].push(...filteredIDs);
+        break;
+      }
+      case "gte": {
+        const filteredIDs = avlGreaterThan(AVLNode, operationValue, true);
+        filtersMap[param].push(...filteredIDs);
+        break;
+      }
+      case "lt": {
+        const filteredIDs = avlLessThan(AVLNode, operationValue, false);
+        filtersMap[param].push(...filteredIDs);
+        break;
+      }
+      case "lte": {
+        const filteredIDs = avlLessThan(AVLNode, operationValue, true);
+        filtersMap[param].push(...filteredIDs);
+        break;
+      }
+      case "eq": {
+        const filteredIDs = avlFind(AVLNode, operationValue) ?? [];
+        filtersMap[param].push(...filteredIDs);
+        break;
+      }
+      case "between": {
+        const [min, max] = operationValue as number[];
+        const filteredIDs = avlRangeSearch(AVLNode, min, max);
+        filtersMap[param].push(...filteredIDs);
+      }
+    }
+  }
+
+  // AND operation: calculate the intersection between all the IDs in filterMap
+  const result = intersect(Object.values(filtersMap)) as unknown as string[];
+
+  return result;
+}
+
+function getSearchableProperties(index: Index): string[] {
+  return index.searchableProperties;
+}
+
+function getSearchablePropertiesWithTypes(index: Index): Record<string, "string" | "number" | "boolean"> {
+  return index.searchablePropertiesWithTypes;
+}
+
+function load(raw: unknown): Index {
+  const {
+    indexes,
+    searchableProperties,
+    searchablePropertiesWithTypes,
+    frequencies,
+    tokenOccurrencies,
+    avgFieldLength,
+    fieldLengths,
+  } = raw as Index;
+
+  return {
+    indexes,
+    searchableProperties,
+    searchablePropertiesWithTypes,
+    frequencies,
+    tokenOccurrencies,
+    avgFieldLength,
+    fieldLengths,
+  };
+}
+
+function save(index: Index): unknown {
```

Comment:
Why are we returning `unknown` here?

Reply:
Are we good on this? Can I resolve this?

Reply:
yes